sktree/tree/_honest_tree.py

```diff
@@ -0,0 +1,338 @@
+            Random Forests", Annals of Statistics, 2019.
+
+    Examples
+    --------
+    >>> from sklearn.datasets import load_iris
+    >>> from sklearn.model_selection import cross_val_score
+    >>> from honest_forests import HonestTreeClassifier
+    >>> clf = HonestTreeClassifier(random_state=0)
+    >>> iris = load_iris()
+    >>> cross_val_score(clf, iris.data, iris.target, cv=10)
+    ...                             # doctest: +SKIP
+    ...
+    array([0.93333333, 0.93333333, 1.        , 1.        , 0.93333333,
+           0.8       , 0.8       , 0.93333333, 1.        , 1.        ])
+    """
+
+    def __init__(
+        self,
+        estimator=None,
+        honest_fraction=0.5,
+        honest_prior="empirical",
+    ):
+        if not estimator:
+            self.estimator_ = DecisionTreeClassifier()
+        self.estimator_ = estimator
+        self.honest_fraction = honest_fraction
+        self.honest_prior = honest_prior
+
+    def fit(self, X, y, sample_weight=None, check_input=True):
+        """Build an honest tree classifier from the training set (X, y).
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            The training input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csc_matrix``.
+
+        y : array-like of shape (n_samples,) or (n_samples, n_outputs)
+            The target values (class labels) as integers or strings.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            Sample weights. If None, then samples are equally weighted. Splits
+            that would create child nodes with net zero or negative weight are
+            ignored while searching for a split in each node. Splits are also
+            ignored if they would result in any single class carrying a
+            negative weight in either child node.
+
+        check_input : bool, default=True
+            Allow to bypass several input checking.
+            Don't use this parameter unless you know what you do.
+
+        Returns
+        -------
+        self : HonestTreeClassifier
+            Fitted estimator.
+        """
+        if check_input:
+            X, y = check_X_y(X, y)
+        self._inherit_estimator_attributes()
+
+        # Account for bootstrapping too
+        if sample_weight is None:
+            sample_weight = np.ones((X.shape[0],), dtype=np.float64)
+
+        nonzero_indices = np.where(sample_weight > 0)[0]
+
+        self.structure_indices_ = np.random.choice(
+            nonzero_indices,
+            int((1 - self.honest_fraction) * len(nonzero_indices)),
+            replace=False,
+        )
+        self.honest_indices_ = np.setdiff1d(nonzero_indices, self.structure_indices_)
+
+        sample_weight[self.honest_indices_] = 0
+
+        # Learn structure on subsample
+        self.estimator_.fit(
+            X,
```

Comment:
Should this be `X[self.structure_indices_,:]` and `y[self.structure_indices_, :]`?